lib/puppetserver/ca/list_action.rb

```diff
@@ -0,0 +1,145 @@
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--config CONF', 'Custom path to Puppet\'s config file') do |conf|
+            parsed['config'] = conf
+          end
+          opts.on('--help', 'Display this command specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--all', 'List all certificates') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        all_certs = get_all_certs(puppet.settings)
+        return 1 if all_certs.nil?
+
+        requested, signed, revoked = separate_certs(all_certs)
+        input['all'] ? human_readable(requested, signed, revoked) : human_readable(requested)
+
+        return 0
+      end
+
+      def human_readable(requested, signed = [], revoked = [])
+
+        if revoked.empty? && signed.empty? && requested.empty?
+          @logger.inform "No certificates to list"
+          return
+        end
+
+        unless requested.empty?
+          @logger.inform "Requested Certificates:"
+          print_output(requested)
+        end
+
+        unless signed.empty?
+          @logger.inform "Signed Certificates:"
+          print_output(signed)
+        end
+
+        unless revoked.empty?
+          @logger.inform "Revoked Certificates:"
+          print_output(revoked)
+        end
+      end
+
+      def print_output(certs)
+        padded = 0
+        certs.each do |cert|
+          cert_size = cert["name"].size
+          padded = cert_size if cert_size > padded
+        end
+
+        certs.each do |cert|
+          @logger.inform "    #{cert["name"]}".ljust(padded + 6) + " (SHA256) #{cert["fingerprints"]["SHA256"]}" +
+                             (cert["dns_alt_names"].empty? ? "" : "\talt names: #{cert["dns_alt_names"]}")
+          end
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def separate_certs(all_certs)
+        requested = all_certs.find_all { |v| v["state"] == "requested" }
+        signed    = all_certs.find_all { |v| v["state"] == "signed"    }
+        revoked   = all_certs.find_all { |v| v["state"] == "revoked"   }
+        return requested, signed, revoked
+      end
+
+      def get_all_certs(settings)
+        result = get_certificate_statuses(settings)
+
+        unless result.is_a? Net::HTTPOK
```

Comment:
Can we update `HttpUtilities::Connection#get` to return an `HttpUtilities::Result` and then check that the code is '200' rather than the class is `Net::HTTPOK`?

Reply:
Yep, and then I'll modify the sign action to support that.